### PR TITLE
[swift-build--init] Consider executing `--init` twice.

### PR DESCRIPTION
**Pull request by JPMartha**
This is a proposal for `--init` behavior.

I'm sorry if I couldn't explain it well.

### Summary

- If the files and directories created by `--init` do NOT exist, SwiftPM creates them again.
- If they exist, SwiftPM does NOT create/overwrite them. (the same as the current implementation)

===

I suppose it's better to consider the following situation.
Concretely, it's better to create `main.swift` at 4.

For example:

1. Run `$ swift build --init=executable`. (Make a new start.)
2. For one reason or another, remove `Package.swift` and `main.swift`. (e.g. Execute discard changes)
3. Run `$ swift build --init=executable` again.
4. Currently, SwiftPM creates `Package.swift` only. `main.swift` is not created. The reason is `Sources/` directory exists.

===

In other words:

The behavior in **Case1** is different from **Case2**.
I suppose it's better to make them the same behavior.
Concretely, it's better to make them the same as **Case2**.

First, run `$ swift build --init=executable`.
Next,
**Case1:** remove `main.swift`
- Run `$ swift build --init=executable` again.
- Then, an error occurs and SwiftPM doesn't execute `--init`.

**Case2:** remove `Package.swift`
- Run `$ swift build --init=executable` again.
- Then, SwiftPM executes `--init` without errors.

===

For your information, when running `$ swift build --init=executable` twice, the following error occurs.
```
error: a manifest file already exists in this directory
```

### CODE REVIEW

**Comment by mxcl** [comment]
I'm not sure `--init` should do anything if the directory has files it would create or overwrite. It makes the command dangerous. /cc @ddunbar

**Comment by ddunbar** [comment]
Yeah, I agree. It would be awesome in the long run to have interactive tools for updating the `Package.swift`, but it requires a lot of infrastructure to be careful about those problems.

I'm not sure the problem this solves (unintentional file removal?) motives the risk here... is there a reason you thought this would commonly come up?

**Comment by aciidb0mb3r** [comment]
I've once run `--init` again by mistake, it really should be not allowed to run if any file is present.

**Comment by JPMartha** [comment]
The point what I wanted to say is `--init` behavior is different according to the target directory condition. I think we should make it the same behavior in any conditions.
Running `--init` twice is not a point.

I said "SwiftPM executes `--init` without errors.", but it is better SwiftPM doesn't allow us to run `--init` twice.

> is there a reason you thought this would commonly come up?

Yes. In fact, I had removed `main.swift` by mistake, so I tried to run `--init` twice to make a fresh start.

Now, as mentioned above, I think it is better SwiftPM doesn't allow us to run `--init` twice.

I'm sorry if I couldn't explain it well.

**Comment by kostiakoval** [comment]
The package is valid if there is a `Package.swift` file, everything else is optional. 
That why if you remove `main.swift` and run `--init` again, you get `a manifest file already exists`
and if you remove `Package.swift` and run  `--init` only  `Package.swift` will be generated.

I feel this is correct behavior. If you do any modification and want to restore them, you should use source-control for that and not Package Manager

**Comment by JPMartha** [comment]
To make this issue clear, I reproduced the same result.

```
$ swift build --init=executable

Creating executable package: MyPackage
Creating Package.swift
Creating .gitignore
Creating Sources/
Creating Sources/main.swift
Creating Tests/
```

```
$ ls -1R

Package.swift
Sources
Tests

./Sources:
main.swift

./Tests:
```

```
$ git status -u

On branch master

Initial commit

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.gitignore
	Package.swift
	Sources/main.swift
```

execute "Discard changes" against all files on GitHub Desktop.app

```
$ git status
On branch master

Initial commit

nothing to commit (create/copy files and use "git add" to track)
```

```
$ ls -1

Sources
Tests
```

```
$ swift build --init=executable

Creating executable package: MyPackage
Creating Package.swift
Creating .gitignore
```

```
$ git status -u
On branch master

Initial commit

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.gitignore
	Package.swift
```

```
$ ls -1R

Package.swift
Sources
Tests

./Sources:

./Tests:
```

Anyway, I think I made a mistake, I don't think is responsible for SwiftPM.
So, I submited as proposal for assisting user's mistakes.

**Comment by kostiakoval** [comment]
I understood what you mean. When running `--init` we can check not only that `Source` folder exist but also that it's not empty.
But I'm not 100% sure that this is safe and desired solution

File: Sources/swift-build/--init.swift [hunk 1]
```diff
@@ -36,7 +36,8 @@ final class InitPackage {
     private func writeManifestFile() throws {
         let manifest = Path.join(rootd, Manifest.filename)
         guard manifest.exists == false else {
-            throw Error.ManifestAlreadyExists
```
Comment (kostiakoval) [inline Sources/swift-build/--init.swift]: I feel it's better to throw an error here instead of calling return, if Manifest already exists.

**Comment by mxcl** [comment]
I can imagine it being useful to *“SwiftPM-ify”* an existing Xcode project. However that is more than this PR. My current feeling is it should be an error to run `--init` if the directory is not empty, or at least, has anything more than `.git` and some root files like `README`.

**Comment by JPMartha** [comment]
I'm sorry that I couldn't explain it well.
Now, I think it is safe not to do anything if the directory is not empty, in spite of the fact that I submitted this PR.
Also, I think it is risky for the future to change the behavior according to target directory's condition.

**Comment by ddunbar** [comment]
I agree with @mxcl here, this command is most useful to start a new project... running it in any other context is more likely than not to be disappointing or wrong.

**Comment by ddunbar** [comment]
I'm not sure what to do with this, @JPMartha can you raise an SR on bugs.swift.org first if you think there is something that needs to be changed, showing the test case?
